packages/runtime/container-runtime/src/blobManager.ts

```diff
@@ -723,24 +729,54 @@ export class BlobManager extends TypedEventEmitter<IBlobManagerEvents> {
 			return [];
 		}
 
-		// The routes or blob node paths are in the same format as returned in getGCData -
-		// `/<BlobManager.basePath>/<blobId>`.
-		for (const route of sweepReadyBlobRoutes) {
-			const pathParts = route.split("/");
-			assert(
-				pathParts.length === 3 && pathParts[1] === BlobManager.basePath,
-				0x586 /* Invalid blob node id in deleted routes. */,
-			);
-			const blobId = pathParts[2];
+		this.deleteBlobs(sweepReadyBlobRoutes);
+		return Array.from(sweepReadyBlobRoutes);
+	}
+
+	/**
+	 * Delete blobs with the given routes. The routes are GC nodes paths of format `/<BlobManager.basePath>/<blobId>`.
+	 * The blob ids are all local ids. Deleting the blobs involves 2 steps:
+	 * 1. The redirect table entry for the local ids are deleted.
+	 * 2. If the storage ids corresponding to the deleted local ids are not in-use anymore, the redirect table entries
+	 * for the storage ids are deleted as well.
+	 */
+	private deleteBlobs(blobRoutes: string[]) {
+		if (blobRoutes.length === 0) {
+			return;
+		}
+
+		// This tracks the storage ids of local ids that are deleted. After the local ids have been deleted, if any of
+		// these storage ids are unused, they will be deleted as well.
+		const maybeUnusedStorageIds: Set<string> = new Set();
+		for (const route of blobRoutes) {
+			const blobId = getBlobIdFromGCNodePath(route);
 			if (!this.redirectTable.has(blobId)) {
 				this.mc.logger.sendErrorEvent({
 					eventName: "DeletedAttachmentBlobNotFound",
 					blobId,
 				});
+				continue;
 			}
+			const storageId = this.redirectTable.get(blobId);
+			assert(!!storageId, "Must be attached to run GC");
+			maybeUnusedStorageIds.add(storageId);
 			this.redirectTable.delete(blobId);
 		}
-		return Array.from(sweepReadyBlobRoutes);
+
+		// Find out storage ids that are in-use and remove them from maybeUnusedStorageIds. A storage id is in-use if
+		// the redirect table has a local id -> storage id entry for it.
+		for (const [localId, storageId] of this.redirectTable.entries()) {
+			assert(!!storageId, "Must be attached to run GC");
+			// For every storage id, the redirect table has a id -> id entry. These do not make the storage id in-use.
+			if (maybeUnusedStorageIds.has(storageId) && localId !== storageId) {
+				maybeUnusedStorageIds.delete(storageId);
+			}
+		}
+
+		// For unused storage ids, delete their entries from the redirect table.
+		for (const storageId of maybeUnusedStorageIds) {
```

Comment:
I love functional programming so I would write this as `maybeUnusedStorageIds.forEach(this.redirectTable.delete)`.  But it's fine as-is.

Reply:
Never mind, it hits a (wrong) lint error